framework/src/utils/RankTwoTensor.C

```diff
@@ -481,44 +481,48 @@ RankTwoTensor::positiveProjectionEigenDecomposition(std::vector<Real> & eigval,
   // C. Miehe and M. Lambrecht, Commun. Numer. Meth. Engng 2001; 17:337~353
 
   // Compute eigenvectors and eigenvalues of this tensor
-  (*this).symmetricEigenvaluesEigenvectors(eigval, eigvec);
+  this->symmetricEigenvaluesEigenvectors(eigval, eigvec);
 
   // Separate out positive and negative eigen values
   std::array<Real, N> epos;
+  std::array<Real, N> d;
   for (unsigned int i = 0; i < N; ++i)
+  {
     epos[i] = (std::abs(eigval[i]) + eigval[i]) / 2.0;
+    d[i] = eigval[i] > 0 ? 1.0 : 0.0;
+  }
 
   // projection tensor
   RankFourTensor proj_pos;
   RankFourTensor Gab, Gba;
   RankTwoTensor Ma, Mb;
 
-  for (unsigned int i = 0; i < N; ++i)
+  for (unsigned int a = 0; a < N; ++a)
   {
-    Ma.vectorOuterProduct(eigvec.column(i), eigvec.column(i));
-    Mb.vectorOuterProduct(eigvec.column(i), eigvec.column(i));
-    Gab = Ma.outerProduct(Mb);
-    if (eigval[i] > 0.0)
-      proj_pos += Gab;
+    Ma.vectorOuterProduct(eigvec.column(a), eigvec.column(a));
+    proj_pos += d[a] * Ma.outerProduct(Ma);
   }
 
-  for (unsigned int i = 0; i < N; ++i)
-  {
-    for (unsigned int j = 0; j < N && j != i; ++j)
+  for (unsigned int a = 0; a < N; ++a)
+    for (unsigned int b = 0; b < N; ++b)
```

Comment:
the previous loop was basically

```
    for (unsigned int j = 0; j < i; ++j)	
```

which loops only over a triangular matrix part. In 3d thats 3 iterations rather than 6 in your code.

At first glance I'd say your loop body is symmetrical with respect to `a` and `b` except for maybe line 520. Can you comment on why the double amount of iterations is necessary here?

Reply:
I see... I can reduce the number of iterations to 3. There is still some subtle difference in the case of equal eigenvalues:

In the original code:
```
proj_pos += (Gab + Gba) * 0.5;
```
In this code:
```
theta_ab = 0.25 * (d[a] + d[b]);
```

You can see that originally we were always assuming d[a] = d[b] = 1